### PR TITLE
Don't run value during inference

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)
This PR adds two small optimizations to help speed up Torch.

1st: Don't evaluate Critic network during inference. This should speed up Actor-Critic algorithms (especially w/ separate Actor Critic) a bit.

2nd: Do soft update (for SAC) in-place and without gradients. Also move this method to `utils.py`. 

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 1]
```diff
@@ -40,7 +40,11 @@ def __init__(
             else 0
         )
 
-        self.visual_processors, self.vector_processors, encoder_input_size = ModelUtils.create_input_processors(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:43]: Is this reformatting necessary?

**Review by vincentpierre — APPROVED** [review]

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 2]
```diff
@@ -435,20 +435,17 @@ def update(self, batch: AgentBuffer, num_sequences: int) -> Dict[str, float]:
...
-        ) = self.policy.sample_actions(
+        (sampled_actions, log_probs, _, _) = self.policy.sample_actions(
             vec_obs,
             vis_obs,
             masks=act_masks,
             memories=memories,
             seq_len=self.policy.sequence_length,
             all_log_probs=not self.policy.use_continuous_act,
         )
+        sampled_values, _ = self.policy.actor_critic.critic_pass(
```
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:446]: It's the value of the vec_obs and vis_obs sampled from the buffer. Renamed to value_estimates for clarity.

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/networks.py:43]: Weird, it was done by my `black` but got undid when I ran it again. Removed.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 3]
```diff
@@ -202,10 +202,22 @@ def sac_q_loss(
         return q1_loss, q2_loss
 
     def soft_update(self, source: nn.Module, target: nn.Module, tau: float) -> None:
-        for source_param, target_param in zip(source.parameters(), target.parameters()):
-            target_param.data.copy_(
-                target_param.data * (1.0 - tau) + source_param.data * tau
-            )
+        """
+        Performs an in-place polyak update of the target module based on the source,
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:206]: I am actually not sure this is Polyak update. I would say `Exponential moving average update`

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:206]: Mention that a tau = 1 means replace the target network weights with the source.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 4]
```diff
@@ -202,10 +202,22 @@ def sac_q_loss(
         return q1_loss, q2_loss
 
     def soft_update(self, source: nn.Module, target: nn.Module, tau: float) -> None:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:204]: I would make this method static (and even move it to ModelUtils)

**Review by vincentpierre — APPROVED** [review]

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:206]: OpenAI and Stable Baselines refer to it as polyak: 
image

Added the comment about tau

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py]: Moved, and added test
